csrc/scheduler/normalization_inner.cpp

```diff
@@ -252,6 +261,386 @@ int64_t getMaxPersistentBatch(
+  int64_t persistent_buffer_size =
+      buffer_bytes_per_batch * persistent_batch_size;
+  auto reg_occ = getMaxRegisterCountPerThreadAndOccupancy(
+      persistent_buffer_size,
+      threads_per_block,
+      target_warps_per_sm,
+      register_overhead);
+  params.register_per_thread = reg_occ.first;
+  params.occupancy = reg_occ.second;
+  params.register_overhead = params.register_per_thread -
+      persistent_buffer_size / scheduler_utils::bytes_per_register;
+  // (4) Calculate other quantities reflecting the quality of the heuristic.
+  // Reduction elements after vectorization are split into multiple persistent
+  // batches.
+  // (4.1) For each persistent batch, [bdimx] is usually padded to full
+  // warps. The number of padded threads are quantified as [n_threads_tails].
+  // Prefer 0 to avoid divergence of the last warp.
+  params.n_threads_tails = params.padded_bdimx - params.bdimx;
+
+  // (4.2) when [reduction_count_after_vectorize] is not divisible by
+  // [persistent_val], the last batch is not be fully utilized, the wasted
+  // threads in the last batch is quantified as [n_persistent_tails].
+  params.n_persistent_tails =
+      ceilDiv(reduction_count_after_vectorize, persistent_batch_size) *
+          persistent_batch_size -
+      reduction_count_after_vectorize;
+
+  auto iters_per_wave = params.occupancy * device_warp_size /
+      threads_per_block * dev_prop->multiProcessorCount;
+  params.n_waves = ceilDiv(total_iteration_numel, iters_per_wave);
+  return params;
+}
+
+// Return true if ha is better than hb
+
+bool compareTwoHeuristics(
+    const HeuristicParams& ha,
+    const HeuristicParams& hb,
+    const int64_t target_register_overhead,
+    const int64_t target_warps_per_sm,
+    const bool prioritize_occupancy) {
+  auto compare = [](int64_t a, int64_t b) -> int {
+    return a > b ? 1 : (a < b ? -1 : 0);
+  };
+  int score = 0;
+
+  if (prioritize_occupancy) {
+    // prefer occupancy larger than target
+    score = compare(
+        ha.occupancy >= target_warps_per_sm,
+        hb.occupancy >= target_warps_per_sm);
+    if (score != 0) {
+      return score > 0;
+    }
+
+    // prefer reduction count after vectorization is divisible by persistent
+    // batch size
+    score = compare(ha.n_persistent_tails == 0, hb.n_persistent_tails == 0);
+    if (score != 0) {
+      return score > 0;
+    }
+
+    // Large register overhead leaves more registers for the compiler to
+    // optimize generated SASS code and avoids register spills. But don't want
+    // to use a very large block size which increaes inter-thread communication.
+    // This condition influence dropout_layer_norm at 7K and softmax_dropout at
+    // 6K and 16K.
+    constexpr int64_t opt_max_threads_per_block = 512l;
+    score = compare(
+        ha.register_overhead > target_register_overhead &&
+            ha.padded_bdimx <= opt_max_threads_per_block,
+        hb.register_overhead > target_register_overhead &&
```

Comment:
`register_overhead` is it more like `allowed_register_overhead`? so the higher the better?

More importantly, line 397 & 399 should have the same compare operator.

Reply:
yes. `    // Large register overhead leaves more registers for the compiler to optimize generated SASS code and avoids register spills.`

line 397 `ha.register_overhead > target_register_overhead` & 399  `hb.register_overhead > target_register_overhead` have the same compare operator, right? both are `>`. 
Or you mean `ha.padded_bdimx <= opt_max_threads_per_block`? Here these two conditions are merged with `and`. we want to use high register overhead but don't want to achieve it at the cost of using a very large block size.

Maybe we should rename `register_overhead` to `spare_registers` or `free_registers` to more precisly refelecting their purpose. They are specifically set aside for transient or temporary usage rather than storing persistent buffers.

Reply:
oh sorry I must have been blind... the compare operator looks correct here. Somehow I read them wrong earlier.

Reply:
@naoyam this dicussion. we were using `register_overhead`, but the sort algorithm prefers a larger value, then the logic doesn't seem right since we usually want to have a small overhead instead of a large one.

Reply:
Not sure what you mean. It actually doesn't prefer a larger value but just prefers a heuristics that has a larger free registers than the set minimum count. If both heuristics have enough registers, there's no preference made. Am I understanding it correctly?

Reply:
yes. your understanding is correct.